### PR TITLE
fs: littlefs: make device access functions configurable

### PR DESCRIPTION
**Pull request by benediktibk**
This changes checks if the read, prog, erase and sync functions
in the littlefs config have already been set. This allows the usage
of custom devices underneath, for instance a SD card or a
custom block level encryption.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

### CODE REVIEW

**Comment by benediktibk** [comment]
Second try, slightly overhauled. An additional sidenote: I did not add a mechanism to remove the flash operations completely in case one wants to use the custom version, as it might be possible that they coexist in system. Because, ... why not?

**Comment by benediktibk** [comment]
By the way, if somebody has a better idea for the name of the macro "FS_LITTLEFS_DECLARE_CUSTOM_OPERATIONS_CONFIG": Don't hesitate to comment it. I don't like this one anyway, but couldn't come up with something better. And I probably shouldn't change the already existing one "FS_LITTLEFS_DECLARE_CUSTOM_CONFIG" to avoid breaking some other code. Therefore, I feel kinda limited in my options.

**Comment by Laczen** [comment]
> Second try, slightly overhauled. An additional sidenote: I did not add a mechanism to remove the flash operations completely in case one wants to use the custom version, as it might be possible that they coexist in system. Because, ... why not?

@benediktibk, if there is no use of these routines they should not be there because it is a waste of space.

**Comment by benediktibk** [comment]
> > Second try, slightly overhauled. An additional sidenote: I did not add a mechanism to remove the flash operations completely in case one wants to use the custom version, as it might be possible that they coexist in system. Because, ... why not?
> 
> @benediktibk, if there is no use of these routines they should not be there because it is a waste of space.

This now depends on CONFIG_FLASH_MAP. Hopefully the correct indicator?

File: include/fs/littlefs.h
Comment (de-nordic) [inline include/fs/littlefs.h:151]: Maybe redefine this as `FS_LITTLEFS_DECLARE_CUSTOM_CONFIG_AND_OPS`, above `FS_LITTLEFS_DECLARE_CUSTOM_CONFIG`,  and then use it in the `FS_LITTLEFS_DECLARE_CUSTOM_CONFIG` definition but fill the last four params as 0.

File: include/fs/littlefs.h
Comment (de-nordic) [inline include/fs/littlefs.h:52]: Maybe for the purpose define structure like `device_services` that will gather all the callbacks, and add pointer to such structure here. Because you are removing pointer to `flash_area`, adding pointer here will keep the size of `fs_littlefs` structure the same and will allow to reuse `device_services` among different mount points.

Comment (benediktibk) [reply include/fs/littlefs.h]: done

File: include/fs/littlefs.h
Comment (benediktibk) [reply include/fs/littlefs.h]: done

File: include/fs/littlefs.h
Comment (Laczen) [inline include/fs/littlefs.h:20]: Shouldn't the `context` contain area_id and flash_area, `lfs_device_open()` should be general for all devices.

**Comment by Laczen** [comment]
> > > Second try, slightly overhauled. An additional sidenote: I did not add a mechanism to remove the flash operations completely in case one wants to use the custom version, as it might be possible that they coexist in system. Because, ... why not?
> > 
> > 
> > @benediktibk, if there is no use of these routines they should not be there because it is a waste of space.
> 
> This now depends on CONFIG_FLASH_MAP. Hopefully the correct indicator?

@benediktibk `CONFIG_FLASH_MAP` is not a good indicator. By introducing a new "backend" implicity multiple little_fs systems are introduced (as it is possible to select flash and something else at the same time). I am not sure the proposed PR supports multiple little_fs systems.

File: include/fs/littlefs.h
Comment (benediktibk) [reply include/fs/littlefs.h:20]: The context is getting set by lfs_device_open, based upon the area id. In the flash case this is getting set to an instance of struct flash_area. The context later on is handed over to each of these device functions, as it basically contains the necessary information for the other flash operations. As it is a general void* it can be used differently, depending on the underlying device type and methods. I basically stole this idea from the previous implementation and generalized it a bit.
Hence, lfs_device_open cannot be general for all devices, but instead must be specific for the device type.

**Comment by benediktibk** [comment]
> > > > Second try, slightly overhauled. An additional sidenote: I did not add a mechanism to remove the flash operations completely in case one wants to use the custom version, as it might be possible that they coexist in system. Because, ... why not?
> > > 
> > > 
> > > @benediktibk, if there is no use of these routines they should not be there because it is a waste of space.
> > 
> > 
> > This now depends on CONFIG_FLASH_MAP. Hopefully the correct indicator?
> 
> @benediktibk `CONFIG_FLASH_MAP` is not a good indicator, it can be selected for other systems.
> 
> By introducing a new "backend" implicity multiple little_fs systems are introduced (as it is possible to select flash and something else at the same time). I am not sure the proposed PR supports multiple little_fs systems, I am also unsure how the "default" should be handled. I have the feeling we are overlooking something. Do you have an idea how two little_fs systems (one in flash and one on sd) would be supported ?

I came to the idea of using CONFIG_FLASH_MAP, as this is the parameter which pulls in the flash_area functions:
https://github.com/zephyrproject-rtos/zephyr/blob/ebfb23979c8787e78742890e400218952045e8e9/subsys/storage/CMakeLists.txt
https://github.com/zephyrproject-rtos/zephyr/blob/ebfb23979c8787e78742890e400218952045e8e9/subsys/storage/flash_map/flash_map.c
Would you rather prefer a new config option just for littlefs, probably depending on CONFIG_FLASH_MAP and defaulting to y?

The default should probably be flash, as this was how it worked previously. Selecting anything else will most likely break applications currently using this API.

Different littlefs systems in one application could be supported via using for them two different instances of struct fs_littlefs, with different device operations set. I just realized that I broke the tests and I'm gonna try to fix them and add a test case for such a scenario. I hope this will make it then more clear.

**Comment by Laczen** [comment]
> Would you rather prefer a new config option just for littlefs, probably depending on CONFIG_FLASH_MAP and defaulting to y?
>

Leave this for the moment, maybe it is not needed at all: If the macro `FS_LITTLEFS_DECLARE_DEFAULT_CONFIG(storage);` is not called no flash routines routines should be registered. This macro could call another (more general) macro that registers the flash routines. 
 
> The default should probably be flash, as this was how it worked previously. Selecting anything else will most likely break applications currently using this API.
> 
> Different littlefs systems in one application could be supported via using for them two different instances of struct fs_littlefs, with different device operations set. I just realized that I broke the tests and I'm gonna try to fix them and add a test case for such a scenario. I hope this will make it then more clear.

I will wait for your example, it might be easier than what I'm thinking now.

File: tests/subsys/fs/littlefs/src/test_lfs_basic.c
Comment (benediktibk) [inline tests/subsys/fs/littlefs/src/test_lfs_basic.c:524]: @Laczen Usage of a normal flash device and something different (in this case a in memory storage) could look like this.

Comment (Laczen) [reply tests/subsys/fs/littlefs/src/test_lfs_basic.c:524]: @benediktibk, nice work. What would happen (during mount) if storage_dev has no meaning ?

A proposal could be that if a high value is assigned to storage_dev (e.g. 0x8000) flash is not used and also the mount procedure could be changed to reflect this.

Another proposal could be that a change to littlefs_mount() is required - this seems reasonable as probably different backends require different checks for mounting.

Comment (benediktibk) [reply tests/subsys/fs/littlefs/src/test_lfs_basic.c:524]: Thanks.
You had a good point there, I forgot to check the result of device_services->open. As of know the result of an invalid value for storage_dev depends on the behaviour of the implementation of device_services->open. In the case of flash this will result in an error, as flash_area_open will return an error.

Honestly, I did not get what you are aiming for with this special range of values for storage_dev. Why would we want to have a different mount procedure, besides the already possible adaption via device_services?

Comment (Laczen) [reply tests/subsys/fs/littlefs/src/test_lfs_basic.c:524]: > You had a good point there, I forgot to check the result of device_services->open. As of know the result of an invalid value for storage_dev depends on the behaviour of the implementation of device_services->open. In the case of flash this will result in an error, as flash_area_open will return an error.
>
 Could you modify the test so that for the RAM_DISK the value of storage_dev is not set.

> Honestly, I did not get what you are aiming for with this special range of values for storage_dev. Why would we want to have a different mount procedure, besides the already possible adaption via device_services?

I overlooked that you also modified the mount procedure. The basic idea should be OK. However it is not OK to check for 0 and if it is set it to the flash method. There should be some verification of device_services that all routines are available, not asign them. The asignment when 0 is OK for the "configuration variables" in little_fs but not for the routines. You might get a mix of flash/other backend routines.

I think it is possible to remove any configuration option just by using macro's:  the "GP" macro is allready there (I would rename it to `FS_LITTLEFS_DECLARE_CUSTOM_CONFIG_CUSTOM_OPS`, we need the following macro's that will call the "GP" macro:
a. `FS_LITTLEFS_DECLARE_DEFAULT_CONFIG_CUSTOM_OPS` that calls the GP macro with read_size, prog_size, cache_size, lookahead_size all 0,
b. `FS_LITTLEFS_DECLARE_CUSTOM_CONFIG` that calls the GP macro with read_size, ... custom values and sets the device_services to the flash services, and adds the flash api routines,
c. `FS_LITTLEFS_DECLARE_DEFAULT_CONFIG` that calls the macro under b with read_size, ... set to 0.

Comment (benediktibk) [reply tests/subsys/fs/littlefs/src/test_lfs_basic.c:524]: done

**Comment by Laczen** [comment]
@benediktibk, the next remarks are what I would do to finalize the PR.

a. I am not a fan of assigning 0 to routines as "default" value. Some backend might even allow to use NULL as routine. I think the macro should directly assign the routines (this will expose lfs_flash_read, ...), but IMO this is OK. No check should be done. This will also remove any flash routines when not used.

b. To make the solution "cleaner" some renaming might be required: `lfs_flash_read` -> `fs_littlefs_flash_read` (and the same for `prog`, ... this will reduce the chance of hitting the same name when the routines are exposed.

c. Regarding the configuration: add a `CONFIG_LFS_FLASH_BACKEND` that defaults to y and if enabled selects `flash` and `flash_map`. Add all flash related routines under this "config". Add an ASSERT to the "default" macro's to verify CONFIG_LFS_FLASH_BACKEND is enabled. (BTW I think it is no longer allowed to use a `#ifdef CONFIG_LFS_FLASH_BACKEND` in a include file).

**Comment by benediktibk** [comment]
> @benediktibk, the next remarks are what I would do to finalize the PR.
> 
> a. I am not a fan of assigning 0 to routines as "default" value. Some backend might even allow to use NULL as routine. I think the macro should directly assign the routines (this will expose lfs_flash_read, ...), but IMO this is OK. No check should be done. This will also remove any flash routines when not used.
> 
> b. To make the solution "cleaner" some renaming might be required: `lfs_flash_read` -> `fs_littlefs_flash_read` (and the same for `prog`, ...) this will reduce the chance of hitting the same name when the routines are exposed.
> 
> c. Regarding the configuration: add a `CONFIG_LFS_FLASH_BACKEND` that defaults to y and if enabled selects `flash` and `flash_map`. Add all flash related routines under this "config". Add an ASSERT to the "default" macro's to verify CONFIG_LFS_FLASH_BACKEND is enabled. (BTW I think it is no longer allowed to use a `#ifdef CONFIG_LFS_FLASH_BACKEND` in a include file).

done, except for the 'select' part. While reading some documentation I stumbled upon this advice: https://unix.stackexchange.com/questions/117521/what-is-the-difference-between-select-vs-depends-in-the-linux-kernel-kconfig
Therefore I went for a 'depends on'.

File: subsys/fs/littlefs_fs.c
Comment (Laczen) [inline subsys/fs/littlefs_fs.c:657]: Is this function still required?

**Review by Laczen — CHANGES_REQUESTED** [review]

File: subsys/fs/littlefs_fs.c
Comment (Laczen) [inline subsys/fs/littlefs_fs.c:742]: Can't these assert's be replaced with a `BUILD_ASSERT` in the GP macro?

**Review by Laczen — CHANGES_REQUESTED** [review]

File: subsys/fs/littlefs_fs.c
Comment (benediktibk) [reply subsys/fs/littlefs_fs.c:742]: There is a build assert as well in the macro, but this is basically just a sanity check at the end.

File: subsys/fs/littlefs_fs.c
Comment (benediktibk) [reply subsys/fs/littlefs_fs.c:657]: Yes, because previously user might have configured the fs_littlefs manually. In such a case they propably did not set the read, prog, erase and sync functions. Hence, we still have to set the defaults in case no operation at all has been set.

File: subsys/fs/littlefs_fs.c
Comment (Laczen) [reply subsys/fs/littlefs_fs.c:742]: Isn't this just taking up ROM space? The error code for a BUILD_ASSERT is not taking any space.

What is the error that should he catched by evaluating this twice ?

**Comment by Laczen** [comment]
@benediktibk, don't take the "requested changes" to serious, I accidentally hit the request changes, they where mere comments.

File: subsys/fs/littlefs_fs.c
Comment (benediktibk) [reply subsys/fs/littlefs_fs.c:742]: Programming error, which I consider possible all my shuffling around of function pointers. I do not trust myself, basically. And the assertion macros which I have used so far in my life are usually made in a way, that they can be optimized out for a release build. Is this not the case for __ASSERT?

Comment (Laczen) [reply subsys/fs/littlefs_fs.c:742]: Maybe the are optimized out for a release build. There is no reason to check it twice and maybe a third time before calling the routine (I didn't check if this is the case). These checks IMO tend to make the code less readable.

Comment (de-nordic) [reply subsys/fs/littlefs_fs.c:742]: It will not defend you from "shuffling" erase into "fs->cfg.prog".

**Comment by Laczen** [comment]
@benediktibk, doesn't the automount feature (at the bottom of littlefs_fs.c) also require some changes ? I would limit this automount feature to the flash_api for the time being, but use the newly defined macro to create the fs structure.

**Comment by benediktibk** [comment]
> @benediktibk, doesn't the automount feature (at the bottom of littlefs_fs.c) also require some changes ? I would limit this automount feature to the flash_api for the time being, but use the newly defined macro to create the fs structure.

Nice catch, did not see this one.

File: subsys/fs/littlefs_fs.c
Comment (benediktibk) [reply subsys/fs/littlefs_fs.c]: Convincing arguments, it is done and pushed

File: subsys/fs/littlefs_fs.c
Comment (Laczen) [reply subsys/fs/littlefs_fs.c:657]: Assigning a default based upon NULL makes understanding the code difficult and unpredictable. To avoid any problems and catching all errors (even when the function pointers are changed during execution) a method could be to add the read/write/... functions to your device_services (these could be const) and to define fs->cfg->read/... functions inside littlefs_fs.c:
```
.cfg.read = little_fs_read;
...
int little_fs_read(...)
{
  if (!fs.device_services->read) {
    return -ENOTSUP;
  }
  return fs.devices_services->read(...);
}
```

Comment (benediktibk) [reply subsys/fs/littlefs_fs.c:657]: I don't think it is possible to make devices_services const, as its members might already end up not being set. In case somebody created the instance of fs_littlefs manually. But I agree with the rest, it would be more clear to bundle these functions together in one spot.

**Comment by benediktibk** [comment]
Did a rebase to the latest main because some weird newlibc thread tests were failing. Maybe this helps.

**Comment by benediktibk** [comment]
Looks like the cause is #38258.

**Comment by stephanosio** [comment]
> Looks like the cause is #38258.

A workaround has been merged: #38261.

**Comment by Laczen** [comment]
@benediktibk are you on slack ? Would like to discuss something.

**Comment by benediktibk** [comment]
> @benediktibk are you on slack ? Would like to discuss something.

Now I am, benediktibk@gmail.com

File: include/fs/littlefs.h
Comment (Laczen) [inline include/fs/littlefs.h:49]: `context set in`

File: subsys/storage/flash_map/flash_map.c
Comment (Laczen) [inline subsys/storage/flash_map/flash_map.c:187]: Changes to flash area should not be part of this PR.

File: subsys/storage/flash_map/flash_map.c
Comment (Laczen) [inline subsys/storage/flash_map/flash_map.c:254]: You can't return an error code for this routine (`uint8_t`)

Comment (benediktibk) [reply subsys/storage/flash_map/flash_map.c]: done

File: subsys/storage/flash_map/flash_map.c
Comment (benediktibk) [reply subsys/storage/flash_map/flash_map.c]: done

File: include/fs/littlefs.h
Comment (benediktibk) [reply include/fs/littlefs.h]: done

File: subsys/fs/littlefs_fs.c
Comment (Laczen) [inline subsys/fs/littlefs_fs.c:481]: Why is this copy needed (and also in the next routines)? What happens when someone reorganizes `lfs_config`?

Comment (benediktibk) [reply subsys/fs/littlefs_fs.c:481]: The context is used an identifier by the underlying device operations, like the flash area id for flash. Therefore, the device operations are able to set the context to a certain value in the call to fs->lfs_context.device_services.open. We have to use the context now as well, as we have to remember which device operations were used for this specific device. For this purpose I set the context of lfs_config as well, where the actual context of device_services.open is stored as well. For every hand over down to the device operation it is therefore necessary to restore the original context.

Comment (benediktibk) [reply subsys/fs/littlefs_fs.c:481]: A reorganization of lfs_config should only be problematic if context gets removed or renamed, but in such a case the build would fail anyway as this specific member is accessed by name.

Comment (benediktibk) [reply subsys/fs/littlefs_fs.c:481]: I have to reuse it as a consequence of https://github.com/zephyrproject-rtos/zephyr/pull/38029#discussion_r700828352. Otherwise I do not know in which context the LFS-operations are called (from LFS). The context allows me to figure out which device operations should be used for this specific LFS-operation.

**Review by nvlsianpu — COMMENTED** [review]
I'm starting to reviewing this PR. I have feeling like it should be divided into few commits which might help in making changes more understandable (to me).

**Comment by benediktibk** [comment]
> I'm starting to reviewing this PR. I have feeling like it should be divided into few commits which might help in making changes more understandable (to me).

Okay, I will create you a fake, but straightforward, commit history. Probably easier to review than the real zig-zagging anyway.

**Comment by benediktibk** [comment]
@nvlsianpu I did at least untangle the test related commits and such things from the actual changes. I hope this helps a little bit? In this way every step by itself builds successfully. ... and it looks like I have to patch it a little bit further, the checks fail so far.

**Comment by benediktibk** [comment]
Blocked by #38401

**Comment by benediktibk** [comment]
Block resolved, ready to review again.

**Comment by benediktibk** [comment]
The two additional commits are related to problems in case flash is deactivated for littlefs. This scenario caused some build issues.

**Comment by benediktibk** [comment]
This is still open for me. Is there anything else required to get the approvals? Better splitting, different approach to the implementation, ... ?

**Comment by Laczen** [comment]
> This is still open for me. Is there anything else required to get the approvals? Better splitting, different approach to the implementation, ... ?

Hi @benediktibk, I think that in the long run it would be better to provide separation into raw flash (already there) and block devices (using disk access). To get there first the raw flash interface would need to be separated from little_fs (as little_fs_flash) and a little_fs_disk would need to be created.

But this is only my idea, maybe @nvlsianpu has a different view. As @nvlsianpu is the maintainer it is up to him to decide in what direction this should go.

**Comment by nvlsianpu** [comment]
> Hi @benediktibk, I think that in the long run it would be better to provide separation into raw flash (already there) and block devices (using disk access). To get there first the raw flash interface would need to be separated from little_fs (as little_fs_flash) and a little_fs_disk would need to be created.

Yes, I'm thinking that introducing the proper interface for interact with block devices would be nice step. although I'm not sure whether we shouldn't introduce even more generic interface (we don't have jet).

**Comment by benediktibk** [comment]
Reading through some of the fatfs implementation as well it looks to me like currently fatfs can be used only on SD-cards and littlefs only on flash devices. So, your are proposing as a first step to enable a user to use fatfs and littlefs both either on SD-cards and flash devices, and as a second step implement the possibility to provide custom hooks which could for instance do some encryption or decryption, depending on the actual use case?

**Comment by benediktibk** [comment]
The encryption on a block level of the SD card is a requirement of our customer, hence I cannot do it differently. So what I am searching for is a way to extend the interface of Zephyr to hook into the FS access. Something which already works (as seen in the linked PR), but is unfortunately not yet in a state which would be acceptable for an upstream push. Obviously, as we are discussing it.

I would be really glad if somebody could give me a rough idea of what I should be aiming for to get it in an acceptable state. @nvlsianpu A two step process, first with littlefs on sd cards, second with custom hooks, would be better?

**Comment by Laczen** [comment]
> The encryption on a block level of the SD card is a requirement of our customer, hence I cannot do it differently. So what I am searching for is a way to extend the interface of Zephyr to hook into the FS access. Something which already works (as seen in the linked PR), but is unfortunately not yet in a state which would be acceptable for an upstream push. Obviously, as we are discussing it.
> 
> I would be really glad if somebody could give me a rough idea of what I should be aiming for to get it in an acceptable state. @nvlsianpu A two step process, first with littlefs on sd cards, second with custom hooks, would be better?

Although it might be easier to create a direct interface to the SD card, it should be considered as a block device and just like usb flash used through the disk_access interface. If this is note done there will be a lot more work in the long run (e.g. suporting littlefs on usb flash, ...). Maybe it is possible to add decryption/encryption hooks to the disk_access methods, but I wouldn't count on that since it would would break usability of devices over multiple platforms.

**Comment by benediktibk** [comment]
This implementation at least is obsolete, as somebody else went already ahead with the idea of using littlefs on a block device: #40704

**Pull request closed without merging**
